File: weather/bootstrap.py
"""
Bootstrap factory for WeatherSkill.

Builds a fully-configured WeatherSkill instance with all available
providers, formatters, and senders registered based on environment
configuration.
"""

import os
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .skill import WeatherSkill


def build_default_skill(**overrides) -> "WeatherSkill":
    """
    Build a fully-configured WeatherSkill instance.

    Registers all available providers, formatters, and senders
    based on environment variables. Accepts optional overrides
    for testing or custom configuration.

    Args:
        **overrides: Optional overrides for testing:
            - providers: list[WeatherProvider] -- override default providers
            - formatters: dict[str, WeatherFormatter] -- override default formatters
            - senders: dict[str, WeatherSender] -- override default senders

    Returns:
        Configured WeatherSkill instance
    """
    from .skill import WeatherSkill

    if "providers" not in overrides:
        providers = _build_providers()
    else:
        providers = overrides.pop("providers")

    if "formatters" not in overrides:
        formatters = _build_formatters()
    else:
        formatters = overrides.pop("formatters")

    if "senders" not in overrides:
        senders = _build_senders()
    else:
        senders = overrides.pop("senders")

    return WeatherSkill(
        providers=providers,
        formatters=formatters,
        senders=senders,
    )


def _build_providers():
    """Build the default provider list from environment."""
    from .providers.base import WeatherProvider
    from .providers.hko import HKOProvider
    from .providers.sg_nea import SGNEAProvider
    from .providers.jma import JMAProvider
    from .providers.au_bom import BOMProvider
    from .providers.nz_metservice import MetServiceProvider
    from .providers.us_nws import NWSProvider
    from .providers.id_bmkg import BMKGProvider
    from .providers.de_dwd import DWDProvider

    providers: list[WeatherProvider] = [
        HKOProvider(),
        SGNEAProvider(),
        JMAProvider(),
        BOMProvider(),
        MetServiceProvider(),
        NWSProvider(),
        BMKGProvider(),
        DWDProvider(),
    ]

    # Key-required providers (only when env var is set)
    if os.environ.get("CWA_API_KEY"):
        from .providers.tw_cwa import CWAProvider
        providers.append(CWAProvider())

    if os.environ.get("METOFFICE_API_KEY"):
        from .providers.uk_metoffice import UKMetOfficeProvider
        providers.append(UKMetOfficeProvider())

    if os.environ.get("KMA_SERVICE_KEY"):
        from .providers.kr_kma import KMAProvider
        providers.append(KMAProvider())

    if os.environ.get("TMD_API_TOKEN"):
        from .providers.th_tmd import TMDProvider
        providers.append(TMDProvider())

    owm_key = os.environ.get("OPENWEATHERMAP_API_KEY")
    if owm_key:
        from .providers.openweathermap import OpenWeatherMapProvider
        providers.append(OpenWeatherMapProvider(api_key=owm_key))

    return providers


def _build_formatters():
    """Build the default formatter dict."""
    from .formatters.cli_text import CliTextFormatter
    from .formatters.telegram import TelegramFormatter
    from .formatters.whatsapp import WhatsAppFormatter

    return {
        "text": CliTextFormatter(),
        "telegram": TelegramFormatter(),
        "whatsapp": WhatsAppFormatter(),
    }


def _build_senders():
    """Build the default sender dict from environment."""
    senders = {}

    if os.environ.get("TELEGRAM_BOT_TOKEN"):
        from .senders.telegram import TelegramSender
        senders["telegram"] = TelegramSender()

    return senders

File: src/bootstrap.ts
/* eslint-disable @typescript-eslint/naming-convention */
/**
 * Bootstrap factory for the Bun WeatherSkill.
 *
 * Mirrors `weather/bootstrap.py`. Registers all v0.1 batch-1 providers
 * unconditionally and the OpenWeatherMap fallback if the
 * `OPENWEATHERMAP_API_KEY` environment variable is set.
 *
 * Batch-2 providers (CWA, Met Office, BOM, MetService, BMKG, DWD,
 * KMA, TMD) are deferred to PRD-002b and will be wired here when their
 * Bun ports land.
 *
 * Formatters: CLI text, Telegram (MarkdownV2), and WhatsApp are
 * registered unconditionally. The Telegram sender is registered only
 * when `TELEGRAM_BOT_TOKEN` is present in the environment, matching
 * the gating in `weather/bootstrap.py`.
 */

import { CliTextFormatter } from "./formatters/cli_text.js";
import { TelegramFormatter } from "./formatters/telegram.js";
import { WhatsAppFormatter } from "./formatters/whatsapp.js";
import { BOMProvider } from "./providers/au_bom.js";
import { DWDProvider } from "./providers/de_dwd.js";
import { HKOProvider } from "./providers/hko.js";
import { BMKGProvider } from "./providers/id_bmkg.js";
import { JMAProvider } from "./providers/jma.js";
import { KMAProvider } from "./providers/kr_kma.js";
import { MetServiceProvider } from "./providers/nz_metservice.js";
import { OpenWeatherMapProvider } from "./providers/openweathermap.js";
import { SGNEAProvider } from "./providers/sg_nea.js";
import { TMDProvider } from "./providers/th_tmd.js";
import { UKMetOfficeProvider } from "./providers/uk_metoffice.js";
import { NWSProvider } from "./providers/us_nws.js";
import { TelegramSender } from "./senders/telegram.js";
import { WeatherSkill, type WeatherSkillInit } from "./skill.js";
import type {
  IWeatherFormatter,
  IWeatherProvider,
  IWeatherSender,
} from "./types.js";

/**
 * Build a fully-configured WeatherSkill instance.
 *
 * Accepts optional overrides for testing or custom configuration.
 */
export function buildDefaultSkill(
  overrides: WeatherSkillInit = {},
): WeatherSkill {
  return new WeatherSkill({
    providers: overrides.providers ?? buildProviders(),
    formatters: overrides.formatters ?? buildFormatters(),
    senders: overrides.senders ?? buildSenders(),
  });
}

/** Construct the default v0.1 provider list from environment. */
function buildProviders(): IWeatherProvider[] {
  const providers: IWeatherProvider[] = [
    new HKOProvider(),
    new SGNEAProvider(),
    new JMAProvider(),
    new NWSProvider(),
    new DWDProvider(),
    new MetServiceProvider(),
    new BMKGProvider(),
    new BOMProvider(),
  ];

  // Key-required: OpenWeatherMap (global fallback).
  const owmKey = process.env.OPENWEATHERMAP_API_KEY;
  if (owmKey) {
    providers.push(new OpenWeatherMapProvider(owmKey));
  }

  const kmaKey = process.env.KMA_SERVICE_KEY;
  if (kmaKey) {
    providers.push(new KMAProvider(kmaKey));
  }

  const tmdKey = process.env.TMD_API_TOKEN;
  if (tmdKey) {
    providers.push(new TMDProvider(tmdKey));
  }

  const metOfficeKey = process.env.METOFFICE_API_KEY;
  if (metOfficeKey) {
    providers.push(new UKMetOfficeProvider(metOfficeKey));
  }

  return providers;
}

/**
 * Construct the default formatter map. Mirrors
 * `weather/bootstrap.py:_build_formatters`. All three formatters are
 * registered unconditionally — they have no configuration.
 */
function buildFormatters(): Record<string, IWeatherFormatter> {
  return {
    text: new CliTextFormatter(),
    telegram: new TelegramFormatter(),
    whatsapp: new WhatsAppFormatter(),
  };
}

/**
 * Construct the default sender map from environment. Mirrors
 * `weather/bootstrap.py:_build_senders` — Telegram registers only when
 * `TELEGRAM_BOT_TOKEN` is set so `TelegramSender`'s constructor cannot
 * throw during default bootstrap.
 */
function buildSenders(): Record<string, IWeatherSender> {
  const senders: Record<string, IWeatherSender> = {};
  if (process.env.TELEGRAM_BOT_TOKEN) {
    senders["telegram"] = new TelegramSender();
  }
  return senders;
}
